[WIP]Fix/null name fallback

zacck-zz opened the pull request:
# What's in this PR?

In this PR I am implementing a fallback for Users who donate without a name. I changed the reciept email mark up to more friendlier language.

I also changed the way RecieptEmail handles Users when they have a first and when they done



## References
Fixes #819

zacck committed:
Add get_name function to the BaseEmail

joshsmith commented on lib/code_corps/emails/receipt_email.ex:
```diff
@@ -3,9 +3,15 @@ defmodule CodeCorps.Emails.ReceiptEmail do
+  def get_name(%User{ first_name: nil }), do: "there"
+
+  @spec get_name(User.t) :: String.t
+  def get_name(%User{ first_name: name}), do: name
```
I think you can condense the spec to just above the first line and then not separate the lines between these.

joshsmith replied:
I also wonder whether we should place this fn into the `BaseEmail` so that it's available to use across other emails.

zacck-zz replied:
@joshsmith  if the functionality is needed in other emails the function can live in BaseEmail

joshsmith reviewed (commented):
Looks great, just a couple comments.

Also note to self that we will need to take the HTML changes here and propagate them to Postmark when this is merged.

zacck-zz commented:
@joshsmith  so I moved the function to BaseEmail and condensed the typespecs also the tests let me now if this what you are envisioning. How do we go about propagating the changes to postmark?

joshsmith commented on test/lib/code_corps/emails/base_email_test.exs:
```diff
@@ -0,0 +1,10 @@
+  test "get name returns there on nil name" do
+    user = %CodeCorps.User{}
+    assert BaseEmail.get_name(user) == "there"
+  end
```
You'll want to add another test for the `non-nil` case.

joshsmith commented on lib/code_corps/emails/base_email.ex:
```diff
@@ -1,9 +1,16 @@
+
+  @spec get_name(User.t) :: String.t
+  def get_name(%User{ first_name: nil }), do: "there"
+  def get_name(%User{ first_name: name}), do: name
```
Prefer collapsing the spaces in the `struct`, e.g. `%User{first_name: name}` and `%User{first_name: nil}`.

joshsmith commented on lib/code_corps/emails/base_email.ex:
```diff
@@ -1,9 +1,16 @@
+  def get_name(%User{ first_name: nil }), do: "there"
+  def get_name(%User{ first_name: name}), do: name
+
+
```
Would remove these extra two lines.

joshsmith commented on test/lib/code_corps/emails/receipt_email_test.exs:
```diff
@@ -5,10 +5,6 @@ defmodule CodeCorps.Emails.ReceiptEmailTest do
   alias CodeCorps.Emails.ReceiptEmail
 
 
```
Would remove these extra two lines, too.

joshsmith reviewed (changes requested):
Overall logic looks great! Just need one more test and some minor formatting things.

zacck-zz commented:
@joshsmith  pushed for those changes let me know if this is looking better?

joshsmith commented on test/lib/code_corps/emails/base_email_test.exs:
```diff
@@ -3,8 +3,14 @@ defmodule CodeCorps.Emails.BaseEmailTest do
+    f_name = "Zacck"
+    user = %CodeCorps.User{first_name: f_name}
+    assert BaseEmail.get_name(user) == f_name
+  end
```
I just realized too but you can wrap all these in a `describe “get_name/2” do` block and write the test names like “when first_name is nil” and “when first_name is set”.

zacck-zz replied:
Could you show me an example of that? very Keen to do that

joshsmith replied:
Yeah, so the full thing would read:

```elixir
describe "get_name/2" do
  test "returns 'there' when first_name nil" do
    user = %CodeCorps.User{first_name: nil}
    assert BaseEmail.get_name(user) == "there"
  end

  test "returns first_name when first_name present" do
    first_name = "Zacck"
    user = %CodeCorps.User{first_name: first_name}
    assert BaseEmail.get_name(user) == first_name
  end
end
```

Does that help?

joshsmith replied:
@zacck your pushes collapsed this comment, so hopefully you see this.

zacck-zz replied:
ah I see for the grouping of the tests for the functionality I though it was something with  guards

joshsmith replied:
Oh, nope! Nothing fancy. :)

zacck-zz replied:
👍  thats  done 😄

joshsmith commented on lib/code_corps/emails/receipt_email.ex:
```diff
@@ -55,7 +55,7 @@ defmodule CodeCorps.Emails.ReceiptEmail do
       project_url: project |> url(),
       subject: project |> build_subject_line(),
-      user_first_name: charge.user.first_name
+      name: BaseEmail.get_name(charge.user)
```
You probably want to alpha order these for consistency.

joshsmith commented on test/lib/code_corps/emails/base_email_test.exs:
```diff
@@ -0,0 +1,16 @@
+
+  test "get_name returns first_name of user" do
+    f_name = "Zacck"
+    user = %CodeCorps.User{first_name: f_name}
```
You don’t need the intermediate variable here. Can just put the string in directly as the value.

joshsmith replied:
Ignore this. Refer to my example instead.

joshsmith commented on test/lib/code_corps/emails/receipt_email_test.exs:
```diff
@@ -50,7 +50,7 @@ defmodule CodeCorps.Emails.ReceiptEmailTest do
       project_current_donation_goal_description: "Test goal",
       subject: "Your monthly donation to Code Corps",
-      user_first_name: "Jimmy"
+      name: "Jimmy"
```
Alpha order here too.

joshsmith reviewed (commented):
Apologies for back and forth. Couple more things I’d suggest.

joshsmith commented:
Looks good! Rebase and squash to one commit and this is mergeable!

joshsmith commented:
I went ahead and rebased/squashed and fixed a small formatting thing on one of the `@spec`'s. An errant space got in there.

Thanks for the hard work on this!

zacck-zz commented:
Thank you so much for this!!! Away from computers didn't see this yet

On Thursday, December 21, 2017, Josh Smith <notifications@github.com> wrote:

> I went ahead and rebased/squashed and fixed a small formatting thing on
> one of the @spec's. An errant space got in there.
>
> Thanks for the hard work on this!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/code-corps/code-corps-api/pull/1306#issuecomment-353419208>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA2ywzQkvgXEEgMsEch-eTZAhiGjoW9iks5tCp9kgaJpZM4RGeRV>
> .
>


-- 
Regards
Isaac Osiemo - Software Developer
Skype: zacck.os
Tel: +27798259736
